add pyfastcopy to make import/export faster on Linux

lyw07 opened the pull request:
## Summary

This is to use `pyfastcopy` to fix https://github.com/learningequality/kolibri/issues/1211 with the goal that importing from USB and exporting to USB will be faster.

## Reviewer guidance

This PR will only help on speeding up the import from/export to USB on Linux systems. 


## Statistics

Update for pyfastcopy on Ubuntu 16.04:

With Python 2:
The speed for importing from USB increases **16%** with 1GB channel: 60.7129230499s --> 49.8765499592s
The speed for importing from USB increases **80%** with 6GB channel: 2926.13523316s --> 556.304387093s

The speed for exporting to USB increases **39%** with 1GB channel: 82s -->
49.5654168129s
The speed for exporting to USB increases **73%** with 6GB channel: 3367s --> 890.768079996s
    
    
With Python 3:
The speed for importing from USB increases **27%** with 1GB channel: 69.01263284683228s --> 50.56928491592407s
The speed for importing from USB with 6GB channel using `pyfastcopy` is 594.6582343578339s, while without using `pyfastcopy` is 675s at 54% (Didn't finish importing to save some time)

The speed for exporting to USB increases **39%** with 1GB channel: 77s --> 47.392226457595825s
The speed for exporting to USB with 6GB channel using `pyfastcopy` is 706.6006577014923s, while without using `pyfastcopy` is unknown (I didn't test on this one to save some time, but I assume it should be similar to the result I got for Python 2). 

> Note that the actual speed for importing from/exporting to USB should be slower than the numbers listed here, because I used a shared folder on my laptop to act as a USB driver for the VM instead of an actual USB driver, which would write files more slowly. But the relative increase should be similar. 

## Issues addressed

#1211

lyw07 commented:
It turns out that `os.sendfile`, the major implementation in `pyfastcopy`, is not available in mac os python 2 but only in python 3. So thinking about write our only file copy method to increase the speed. Closing this PR now as `pyfastcopy` is not a cross-platform tool.

aronasorman commented on kolibri/content/management/commands/exportcontent.py:
```diff
@@ -47,17 +47,12 @@ def handle_async(self, *args, **options):
-                            file_cp_progress_update(length)
-                        else:
-                            exported_files.append(dest)
+                    size = copy.fast_file_copy()
```
How do we handle file downloads now? I remember that file downloads use the loop above.

aronasorman replied:
Oh sorry, this is content exporting.

lyw07 replied:
Yeah, file download is the same as before.

aronasorman commented on kolibri/content/utils/transfer.py:
```diff
@@ -187,6 +188,14 @@ def __iter__(self):
         return self
 
+    def fast_file_copy(self):
+        # import pdb; pdb.set_trace()
```
Remove PDB here. Can you also add docstrings as to why this is "fast"?

aronasorman commented:
Before I merge, I think you tested this for Mac Linux and windows right @lyw07?

lyw07 commented:
@aronasorman I only tested on Mac and Linux. So on Mac, there's no improvement for the speed of transfer. And on Linux, the improvement is listed above. I haven't tried to test on windows yet. But I can quickly set up a windows VM to test it.

radinamatic commented:
@lyw07 Server not starting from the taskbar (Windows 10, [this installer](https://storage.googleapis.com/le-downloads/kolibri%2Frelease%2F2940%2Freport.html)), and I'm getting this error when trying from Command prompt:

```
C:\Python27\Scripts>kolibri start --debug --foreground
Traceback (most recent call last):
  File "c:\python27\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\python27\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Python27\Scripts\kolibri.exe\__main__.py", line 9, in <module>
  File "c:\python27\lib\site-packages\kolibri\utils\cli.py", line 571, in main
    initialize(debug=debug)
  File "c:\python27\lib\site-packages\kolibri\utils\cli.py", line 140, in initialize
    django.setup()
  File "c:\python27\lib\site-packages\kolibri\dist\django\__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "c:\python27\lib\site-packages\kolibri\dist\django\apps\registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "c:\python27\lib\site-packages\kolibri\dist\django\apps\config.py", line 202, in import_models
    self.models_module = import_module(models_module_name)
  File "c:\python27\lib\importlib\__init__.py", line 37, in import_module
    __import__(name)
  File "c:\python27\lib\site-packages\kolibri\core\models.py", line 5, in <module>
    registry.initialize()
  File "c:\python27\lib\site-packages\kolibri\plugins\registry.py", line 80, in initialize
    plugin_module = importlib.import_module(import_string)
  File "c:\python27\lib\importlib\__init__.py", line 37, in import_module
    __import__(name)
  File "c:\python27\lib\site-packages\kolibri\plugins\coach\kolibri_plugin.py", line 4, in <module>
    from . import hooks, urls
  File "c:\python27\lib\site-packages\kolibri\plugins\coach\urls.py", line 4, in <module>
    from .api_urls import urlpatterns
  File "c:\python27\lib\site-packages\kolibri\plugins\coach\api_urls.py", line 3, in <module>
    from rest_framework import routers
  File "c:\python27\lib\site-packages\kolibri\dist\rest_framework\routers.py", line 25, in <module>
    from rest_framework import views
  File "c:\python27\lib\site-packages\kolibri\dist\rest_framework\views.py", line 18, in <module>
    from rest_framework.request import Request
  File "c:\python27\lib\site-packages\kolibri\dist\rest_framework\request.py", line 22, in <module>
    from rest_framework.settings import api_settings
  File "c:\python27\lib\site-packages\kolibri\dist\rest_framework\settings.py", line 25, in <module>
    from django.test.signals import setting_changed
  File "c:\python27\lib\site-packages\kolibri\dist\django\test\__init__.py", line 27, in <module>
    import mock  # NOQA
  File "c:\python27\lib\site-packages\kolibri\dist\mock\__init__.py", line 2, in <module>
    import mock.mock as _mock
  File "c:\python27\lib\site-packages\kolibri\dist\mock\mock.py", line 71, in <module>
    _v = VersionInfo('mock').semantic_version()
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\version.py", line 461, in semantic_version
    self._semantic = self._get_version_from_pkg_resources()
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\version.py", line 448, in _get_version_from_pkg_resources
    result_string = packaging.get_version(self.package)
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\packaging.py", line 740, in get_version
    version = _get_version_from_git(pre_version)
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\packaging.py", line 665, in _get_version_from_git
    git_dir = git._run_git_functions()
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\git.py", line 131, in _run_git_functions
    if _git_is_installed():
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\git.py", line 83, in _git_is_installed
    _run_shell_command(['git', '--version'])
  File "c:\python27\lib\site-packages\kolibri\dist\pbr\git.py", line 49, in _run_shell_command
    env=newenv)
  File "c:\python27\lib\subprocess.py", line 390, in __init__
    errread, errwrite)
  File "c:\python27\lib\subprocess.py", line 640, in _execute_child
    startupinfo)
TypeError: environment can only contain strings

C:\Python27\Scripts>
```

lyw07 commented:
@radinamatic thank you! Did we have some fix on Windows recently? I haven't pull from release-0.6.x branch for a long time, so I'll do that right now to get a new installer.

radinamatic commented:
Not that I recall @lyw07, the last build I did test on Windows was #2331, and there were no startup errors, not sure if something changed in the meantime.

lyw07 commented:
Seeing the error by doing `import os; os.environ` inside subprocess.py. It shows that one of the environment variable is a unicode instead of a string. The variable is `u'KOLIBRI_HOME' : u'C:\\Users\\IEUser\\.kolibri'`

indirectlylit commented:
> I can quickly set up a windows VM to test it

lies! ;-)

radinamatic commented:
`Windows`, `VM` and `quickly` do not really go together...

aronasorman commented on requirements/cext.txt:
```diff
@@ -1 +1 @@
-cryptography==2.0.3
\ No newline at end of file
+cryptography
```
It's best to pin the exact versions we need, so that it's easier to build this commit when we go back from the future.

lyw07 replied:
The problem is that we use a repository called `piwheels` to get `cryptography` for raspberry pi. `piwheels` only contains `cryptography` with 2.0.3 for now. However, `pypi` now has 2.1 for `cryptography` and some of the 2.0.3 wheels are deleted. So I couldn't get a consistent version of cryptography that works for both `piwheels` and `pypi`. What I do in my code now is to get the latest version of cryptography in both places.

bennuttall replied:
piwheels is keeping up with new package releases quite quickly (we were probably just paused (or working through the backlog) when you found we were missing the latest version of cryptography). We have it now: https://www.piwheels.hostedpi.com/simple/cryptography/

aronasorman commented on install_cexts.py:
```diff
@@ -79,18 +79,28 @@ def install_package_by_wheel(path, name):
     Parse the PYPI and Piwheel link for the package and install the desired wheel files.
     """
 
+    dict = {}
```
This overwrites the `dict` default function. Let's change this variable name.

radinamatic commented:
Still getting the same error on Windows as above, with the latest `exe` from Buildkite.

lyw07 commented:
@aronasorman @indirectlylit @radinamatic Sorry for the late reply! I was trying to change some structure of the previous code, and that took me some time. Sorry about that and thank you for the waiting :)

So the problem I found on my previous code was that `pyfastcopy` uses a C python package called `pysendfile`, which is not cross-platform. The C code compiled on Linux in the builds couldn't be run on other platforms. Although I tested on Mac OS X before, the program I tested was originally built on Mac OS X, so I didn't realize the error until today. This `pyfastcopy` also caused the error that Radina reported above on Windows, as it was trying to import `mock`, but it didn't exist on Windows. 

What I've changed in code is to install `pyfastcopy` into `kolibri/dist/cext` folder and add its path to `sys.path` manually, as it is a C extension just like `cryptography`. However, since `pyfastcopy` is not working on Mac OS X and Windows, I've changed the code so that `pyfastcopy` will only be installed and used on Linux platforms. 

I've tested the pex file on Mac OS X, Ubuntu 16.04 and windows installer on Windows 10, doing import from network, import from/export to USB. The export to USB is working once I manually change the `block_size` to 2097152 as stated in issue #2425. Except that, everything looks good to me.

Please let me know if you find any other errors or any code implementation I can improve. Thank you!!

lyw07 commented:
Looks like `pysendfile` doesn't work on 32-bit systems. I will test on a 32-bit Ubuntu now.

lyw07 commented:
This PR will only benefit linux-x86_64 platform since our builds run on this platform. The limitation of the C extension `pysendfile` makes it only accessible on the platform that compiles it. Therefore, after discussing with @aronasorman, we decide to move this to 0.7 and maintain the general approach we have currently for all the platforms on 0.6. We can handle this `pyfastcopy` the same as the other c extension `cryptography` in the future, probably having different installers for different platforms as Ben mentioned before (Not sure). But for now, closing this PR to reduce the complexity of the code that doesn't benefit the speed of transfer much may be a good idea.
